airdrop-machine/sources/airdrop_machine.move

```diff
@@ -0,0 +1,289 @@
+    }
+
+    public fun mint_impl(
+        _minter: &signer,
+        collection: Object<Collection>,
+        name: String,
+        description: String,
+        uri: String,
+        recipient_addr: address,
+    ): Object<Token> acquires CollectionConfig {
+        assert!(minting_started(collection), error::permission_denied(EMINTING_HAS_NOT_STARTED));
+
+        let collection_signer = collection_signer(collection);
+        
+        let constructor_ref = &token::create_numbered_token(
+            &collection_signer,
+            collection::name(collection),
+            description,
+            name,
+            utf8(b""), // name_with_index_suffix 
+            royalty::get(collection),
+            uri,
+        );
+
+        token_components::create_refs(constructor_ref);
+        transfer_token::transfer(&collection_signer, recipient_addr, constructor_ref);
+        object::object_from_constructor_ref(constructor_ref)
+    }
+
+    public fun create_collection_impl(
+        admin: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        mutable_collection_metadata: bool,
+        mutable_token_metadata: bool,
+        tokens_burnable_by_collection_owner: bool,
+        tokens_transferrable_by_collection_owner: bool,
+        max_supply: Option<u64>,
+        royalty_numerator: Option<u64>,
+        royalty_denominator: Option<u64>,
+    ): Object<Collection> {
+        let admin_addr = signer::address_of(admin);
+        let object_constructor_ref = &object::create_object(admin_addr);
+        let object_signer = object::generate_signer(object_constructor_ref);
+        let royalty = royalty(&mut royalty_numerator, &mut royalty_denominator, admin_addr);
+
+        let constructor_ref = if (option::is_some(&max_supply)) {
+            collection::create_fixed_collection(
+                &object_signer,
+                description,
+                option::extract(&mut max_supply),
+                name,
+                royalty,
+                uri,
+            )
+        } else {
+            collection::create_unlimited_collection(
+                &object_signer,
+                description,
+                name,
+                royalty,
+                uri,
+            )
+        };
+
+        collection_components::create_refs_and_properties(&constructor_ref);
+        let collection = object::object_from_constructor_ref(&constructor_ref);
+        configure_collection_and_token_properties(
+            &object_signer,
+            collection,
+            mutable_collection_metadata,
+            mutable_token_metadata,
+            tokens_burnable_by_collection_owner,
+            tokens_transferrable_by_collection_owner,
+        );
+
+        let object_signer = object::generate_signer(&constructor_ref);
+
+        move_to(&object_signer, CollectionConfig {
+            extend_ref: object::generate_extend_ref(object_constructor_ref),
```

Comment:
This is the wrong constructor ref.

Reply:
I dont need the collections construct ref but rather object signer, as this is the one we need to get the signer which creates the collection to create the token during minting not the collection signer.

Reply:
Ah forget this, was confused! So we are sending the extend ref of the creator object to the collection signer.

The CollectionConfig which contains the Collection owners ExtendRef shouldn't be in the collection, but in the Collection owner object. Something like

```
    inline fun get_collection_owner_signer(): &signer {
        let address = object::create_object_address(&@deploy_addr, OBJECT_SEED);
        let config = borrow_global<CollectionConfig>(address);
        &object::generate_signer_for_extending(&config.extend_ref)
    }
```

Reply:
We redefine `object_signer` to be the collection object signer, which isn't what we want. `extend_ref` here is the `signer` for the collection owner, so we want this resource to be sent to the collection owner object. Hope that's clear :)

Reply:
  ❤️❤️